File: packages/music-player/android-app/app/src/main/kotlin/dev/monochromatic/musicplayer/core/RelPath.kt
// Summary (folded in from the original KDoc on this file):
//   Pure relative-path display. We strip the longest common directory prefix from a queue's track
//   paths so the UI shows each track relative to the loaded root (`Artist/Album/01.flac` instead of
//   the full path, or a bare filename when the whole queue is one folder). This is a faithful port of
//   the desktop player's `relpath.rs`; it has no platform dependency and is fully unit-tested against
//   the same Rust test vectors. The pagination grouping consumes this output, and the fallback to the
//   full path guards a degenerate input that would otherwise collapse a UI row to an empty label.

// What:     `package dev.monochromatic.musicplayer.core` names the namespace this file's
//           declarations live in. Every Kotlin file starts with a `package` line; the dotted name
//           mirrors the directory path `dev/monochromatic/musicplayer/core/`. Anything declared here
//           (the `SEPARATOR` constant, the three functions) is reachable from other files in the same
//           package without an import, and from other packages by importing this name.
// Why:      We need the namespace so the rest of the app (the queue, the pagination code) can refer to
//           `relativeDisplayPaths` as a member of `dev.monochromatic.musicplayer.core`.
//
// In TS you'd write (pseudocode):
// ```ts
// // No 1:1 equivalent — in TS the module identity comes from the file path, there is no
// // `package` statement. Picture this file living at src/core/RelPath.ts and being imported as
// // `import { relativeDisplayPaths } from "./core/RelPath";`.
// ```
package dev.monochromatic.musicplayer.core

// What:     `private const val SEPARATOR = "/"` declares a top-level constant.
//           - `private` limits visibility to THIS file (other files cannot see `SEPARATOR`).
//           - `const val` is a COMPILE-TIME constant: its value is baked into the bytecode wherever
//             it is used, so it must be a primitive or `String` literal known at compile time.
//             Siblings the reader might have expected:
//               * plain `val` — a read-only binding computed at runtime (can hold any object, not
//                 inlined). `const val` is the stricter, compile-time-only form.
//               * `var` — a re-assignable (mutable) binding. We do not want that here.
//           - The type is inferred as `String` from the `"/"` literal; we did not write `: String`.
// Why:      The queue is path-based and the UI/pagination expect a single `/` separator; naming it
//           once avoids a bare `"/"` literal scattered through the split and the re-join below.
//
// In TS you'd write (pseudocode):
// ```ts
// const SEPARATOR = "/";
// ```
/**
 * Defines separator value for this music-player component; the TypeScript-oriented notes above explain its
 * source and use.
 */
private const val SEPARATOR = "/"

// What:     `private fun normalComponents(path: String): List<String>` declares a function.
//           - `private` keeps it visible only inside this file (an internal helper).
//           - `fun` is the function keyword.
//           - `path: String` is the single parameter: an owned, immutable text string. (Kotlin
//             `String` is a reference type but is immutable, so there is no borrowed-vs-owned
//             question the way Rust has `&str` vs `String`.)
//           - `: List<String>` is the return type: a READ-ONLY list of strings. Siblings the reader
//             might have expected:
//               * `MutableList<String>` — the same data but with `add`/`remove` methods exposed.
//               * `Array<String>` — a fixed-size, mutable-element JVM array.
//             We return the read-only `List` interface so callers cannot mutate what we hand back.
// Why:      We need to split a path into just its named folder/file segments, dropping the root and
//           any `.`/`..` markers, so the prefix comparison and re-joining below are clean.
//
// In TS you'd write (pseudocode):
// ```ts
// function normalComponents(path: string): string[] {
//   return path.split("/").filter((seg) => seg !== "" && seg !== "." && seg !== "..");
// }
// ```
/**
 * Defines normal components behavior for this music-player component; the TypeScript-oriented notes above
 * explain its call shape and effects.
 */
private fun normalComponents(path: String): List<String> =
    // What:     The `=` after the signature is Kotlin's EXPRESSION-BODY function form: instead of a
    //           `{ ... }` block with a `return`, the whole function IS the single expression on the
    //           right, and its value is returned implicitly. So this entire `path.split(...).filter(...)`
    //           line is both the body and the return value.
    //           Piece by piece:
    //           - `path.split(SEPARATOR)` cuts the string at every `/` and returns a `List<String>`
    //             of the pieces (including empty pieces for leading/trailing or doubled slashes).
    //           - `.filter { ... }` keeps only the elements for which the lambda returns `true`,
    //             producing a new list. The `{ ... }` is a TRAILING LAMBDA: when a lambda is the last
    //             argument, Kotlin lets you move it outside the parentheses (here `filter`'s only
    //             argument is the lambda, so the parentheses vanish entirely).
    //           - Inside the lambda, `it` is the IMPLICIT single parameter name Kotlin gives a lambda
    //             that takes exactly one argument; here `it` is each segment string. `it.isNotEmpty()`
    //             drops empty pieces; `it != "."` and `it != ".."` drop the relative-path markers.
    // Why:      Dropping empties and `.`/`..` leaves just the real folder and file names, so two paths
    //           can be compared segment-by-segment and rejoined consistently.
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // return path.split("/").filter((seg) => seg !== "" && seg !== "." && seg !== "..");
    // ```
    path.split(SEPARATOR).filter { it.isNotEmpty() && it != "." && it != ".." }

// What:     `private fun commonPrefixLen(lists: List<List<String>>): Int` declares a function.
//           - `private` — file-internal helper.
//           - `lists: List<List<String>>` is the parameter: a read-only list whose elements are
//             themselves read-only lists of strings (one segment-list per track).
//           - `: Int` is the return type: a 32-bit signed integer. Sibling the reader might have
//             expected: `Long` (a 64-bit signed integer). We use `Int` because collection sizes and
//             indices in Kotlin are `Int` (`List.size` is `Int`, `list[i]` indexes with `Int`); using
//             `Long` would force `.toInt()`/`.toLong()` conversions at every index and is pointless
//             for in-memory list lengths.
// Why:      We need to know how many LEADING segments every track shares (capped so the filename
//           always survives); that shared run is the "loaded root" we strip off.
//
// In TS you'd write (pseudocode):
// ```ts
// function commonPrefixLen(lists: string[][]): number { ... }
// ```
/**
 * Defines common prefix len behavior for this music-player component; the TypeScript-oriented notes above
 * explain its call shape and effects.
 */
private fun commonPrefixLen(lists: List<List<String>>): Int {
    // What:     `val shortest = lists.minOfOrNull { it.size } ?: 0`.
    //           - `val` declares a read-only local binding (cannot be reassigned). Sibling: `var`
    //             would be reassignable; we do not need that.
    //           - `lists.minOfOrNull { it.size }` walks every inner list, takes its `.size` (an `Int`),
    //             and returns the SMALLEST size. The trailing lambda `{ it.size }` is the selector;
    //             `it` is each inner list. The `OrNull` suffix means: if `lists` is EMPTY there is no
    //             minimum, so it returns `null` instead of throwing. Its result type is therefore
    //             `Int?` (a nullable `Int`).
    //           - `?: 0` is the ELVIS operator: "if the left side is `null`, use `0` instead". So
    //             `shortest` ends up a plain non-null `Int`.
    // Why:      The common prefix can never be longer than the shortest path's segment count; and an
    //           empty queue must fall back to `0` rather than crash on a `null`.
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // const sizes = lists.map((l) => l.length);
    // const shortest = sizes.length === 0 ? 0 : Math.min(...sizes);
    // ```
    /**
     * Defines shortest value for this music-player component; the TypeScript-oriented notes above explain its
     * source and use.
     */
    val shortest = lists.minOfOrNull { it.size } ?: 0
    // What:     `if (shortest == 0) return 0`. An early `return` when some path has no named segments
    //           at all (so there is nothing to strip). `==` in Kotlin is STRUCTURAL equality (it calls
    //           `.equals()`), which for `Int` is just numeric equality.
    // Why:      Guards the `shortest - 1` line below: if we let `shortest` be 0 the cap would be -1 and
    //           the loop bounds would be nonsense. (In the Rust twin this also avoids unsigned
    //           underflow; in Kotlin `Int` is signed so it would not panic, but the early return is the
    //           same clean guard.)
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // if (shortest === 0) return 0;
    // ```
    if (shortest == 0) return 0
    // What:     `val cap = shortest - 1`. A read-only `Int` holding the most segments we are allowed to
    //           strip, leaving at least the final one (the filename) on every track. `- 1` is ordinary
    //           integer subtraction, identical to TS.
    // Why:      A UI row must never collapse to an empty label, so we always keep at least one segment.
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // const cap = shortest - 1;
    // ```
    /**
     * Defines cap value for this music-player component; the TypeScript-oriented notes above explain its source
     * and use.
     */
    val cap = shortest - 1
    // What:     `var run = 0`. A MUTABLE local `Int`, the counter for how many leading segments match
    //           so far. We use `var` (not `val`) precisely because the loop below reassigns it.
    // Why:      We need a running count we can increment as we confirm each shared segment.
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // let run = 0;
    // ```
    /**
     * Defines run value for this music-player component; the TypeScript-oriented notes above explain its source
     * and use.
     */
    var run = 0
    // What:     `while (run < cap && lists.all { it[run] == lists[0][run] }) { ... }`.
    //           - The loop continues while BOTH conditions hold (`&&` short-circuits, so the
    //             right side is only evaluated while `run < cap`).
    //           - `lists.all { ... }` returns `true` only when the trailing-lambda predicate holds for
    //             EVERY element. `it` is each inner list; `it[run]` indexes that list at position
    //             `run` (safe here because `run < cap < shortest <= every list's size`).
    //           - `lists[0]` is the first track's segment list, used as the reference; `lists[0][run]`
    //             is its segment at position `run`. `==` is structural string equality.
    // Why:      One linear scan finds the longest shared leading run without recursion: as soon as any
    //           track disagrees at position `run`, `.all` is `false` and we stop.
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // while (run < cap && lists.every((l) => l[run] === lists[0][run])) {
    //   run++;
    // }
    // ```
    while (run < cap && lists.all { it[run] == lists[0][run] }) {
        // What:     `run++`. Post-increment: advance the counter past a segment that all tracks share.
        //           This is the one mutation that justifies `run` being a `var`. Character-identical to
        //           TS.
        // Why:      Move on to compare the next position.
        //
        // In TS you'd write (pseudocode):
        // ```ts
        // run++;
        // ```
        run++
    }
    // What:     `return run`. Hand back the shared-prefix length. Because this function has a `{ }`
    //           block body (not the expression-body `=` form used above), the return is EXPLICIT.
    // Why:      The caller needs to know how many leading segments to drop from every track.
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // return run;
    // ```
    return run
}

// What:     `fun relativeDisplayPaths(tracks: List<String>): List<String>` declares the one PUBLIC
//           function of this file (no `private`, so it is visible to the rest of the module — Kotlin's
//           default visibility is `public`).
//           - `tracks: List<String>` is the parameter: a read-only list of full track-path strings, in
//             load order. Sibling the reader might expect: `MutableList<String>`; we take the read-only
//             `List` because we never mutate the input.
//           - `: List<String>` returns one relative display string per track, in the same order.
// Why:      The UI shows folders, not just bare filenames, but the shared absolute prefix (e.g. a long
//           music-library root) is noise; this strips it once per queue.
//
// In TS you'd write (pseudocode):
// ```ts
// function relativeDisplayPaths(tracks: string[]): string[] {
//   if (tracks.length === 0) return [];
//   const componentLists = tracks.map(normalComponents);
//   const prefixLen = commonPrefixLen(componentLists);
//   return componentLists.map((list, i) => {
//     const relative = list.slice(prefixLen).join("/");
//     return relative === "" ? tracks[i] : relative;
//   });
// }
// ```
/**
 * Defines relative display paths behavior for this music-player component; the TypeScript-oriented notes above
 * explain its call shape and effects.
 */
fun relativeDisplayPaths(tracks: List<String>): List<String> {
    // What:     `if (tracks.isEmpty()) return emptyList()`. An early `return` for an empty queue.
    //           - `tracks.isEmpty()` is `true` when the list has no elements.
    //           - `emptyList()` is a factory function that returns a shared, immutable empty
    //             `List<String>`. Siblings the reader might have expected:
    //               * `listOf()` — also makes an immutable list, but `emptyList()` is the canonical
    //                 zero-element form.
    //               * `mutableListOf()` — would make a NEW empty MUTABLE list each call.
    //             We use `emptyList()` because we are returning, not building up, and it avoids an
    //             allocation (the empty list is a singleton).
    // Why:      Nothing to relativize, and `commonPrefixLen` of nothing has no meaningful prefix.
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // if (tracks.length === 0) return [];
    // ```
    if (tracks.isEmpty()) return emptyList()
    // What:     `val componentLists = tracks.map { normalComponents(it) }`.
    //           - `val` — read-only local; its inferred type is `List<List<String>>`.
    //           - `tracks.map { ... }` transforms each element into a new value, producing a new list
    //             of the same length. The `{ ... }` is a trailing lambda; `it` is each track path
    //             string, and we pass it to `normalComponents` to get that track's segment list.
    // Why:      Compute each path's named segments ONCE here, then reuse them both for the shared
    //           prefix and for each track's remainder below.
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // const componentLists = tracks.map((it) => normalComponents(it));
    // ```
    /**
     * Defines component lists value for this music-player component; the TypeScript-oriented notes above explain
     * its source and use.
     */
    val componentLists = tracks.map { normalComponents(it) }
    // What:     `val prefixLen = commonPrefixLen(componentLists)`. A read-only `Int`: how many leading
    //           segments are the shared root. Plain function call, no special syntax; Kotlin passes the
    //           list by reference (there is no borrow annotation as in Rust).
    // Why:      Decide how many leading segments to strip from every track.
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // const prefixLen = commonPrefixLen(componentLists);
    // ```
    /**
     * Defines prefix len value for this music-player component; the TypeScript-oriented notes above explain its
     * source and use.
     */
    val prefixLen = commonPrefixLen(componentLists)
    // What:     `return componentLists.zip(tracks).map { (list, path) -> ... }`. The final, returned
    //           expression, built from two chained calls:
    //           - `componentLists.zip(tracks)` pairs up the two equal-length lists element-by-element,
    //             producing a `List<Pair<List<String>, String>>`: each entry is (this track's segment
    //             list, this track's original full path). We keep the original path around so the
    //             empty-result fallback can reach it.
    //           - `.map { (list, path) -> ... }` transforms each pair into one display string. The
    //             `(list, path)` inside the lambda is a DESTRUCTURING declaration: it unpacks the
    //             `Pair` into two named locals (`list` = the segment list, `path` = the original full
    //             path) instead of using the implicit single `it`.
    // Why:      Produce one relative display string per track while preserving load order.
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // return componentLists.map((list, i) => {
    //   const path = tracks[i];
    //   const relative = list.slice(prefixLen).join("/");
    //   return relative === "" ? path : relative;
    // });
    // ```
    return componentLists.zip(tracks).map { (list, path) ->
        // What:     `val relative = list.drop(prefixLen).joinToString(SEPARATOR)`.
        //           - `list.drop(prefixLen)` returns a NEW list with the first `prefixLen` elements
        //             removed — i.e. the segments that come AFTER the shared root. (It drops a COUNT
        //             from the front; it is not a start..end range slice.)
        //           - `.joinToString(SEPARATOR)` glues the remaining segments back together with `/`
        //             between them, yielding a single `String`.
        // Why:      The segments past the common root ARE the relative path we want to show.
        //
        // In TS you'd write (pseudocode):
        // ```ts
        // const relative = list.slice(prefixLen).join("/");
        // ```
        /**
         * Defines relative value for this music-player component; the TypeScript-oriented notes above explain
         * its source and use.
         */
        val relative = list.drop(prefixLen).joinToString(SEPARATOR)
        // What:     `if (relative.isEmpty()) path else relative`. This is an IF-EXPRESSION: in Kotlin
        //           `if/else` produces a value (like a ternary), so this whole line evaluates to one of
        //           the two branches and, being the last expression in the lambda, becomes the lambda's
        //           result for this element. `relative.isEmpty()` is `true` for the empty string.
        //           NOTE: both branches are already plain `String`s — `path` is the untouched original
        //           input string — so there is no conversion or copying step here (unlike the Rust
        //           twin, whose paths needed a lossy-UTF-8 `to_string_lossy()`; that concern does not
        //           exist in this Kotlin file).
        // Why:      Defensive fallback: a pathological path with no named segments would relativize to
        //           the empty string, which would render as a blank UI row; show its full original text
        //           instead.
        //
        // In TS you'd write (pseudocode):
        // ```ts
        // return relative === "" ? path : relative;
        // ```
        if (relative.isEmpty()) path else relative
    }
}

File: packages/music-player/desktop-app/src/relpath.rs
//! Pure relative-path display: strip the longest common directory prefix from
//! the queue's track paths so the UI can show each track's path relative to the
//! loaded root (e.g. `Artist/Album/01.flac` rather than the full absolute path,
//! or just `01.flac` when the whole queue lives in one folder). No Slint, audio,
//! or I/O, so it is fully unit-tested; `queue::Queue::display_paths` calls this
//! to build the list handed to the UI, and `pagination` then groups the result
//! (by folder for subfolder tracks, by first letter for root-level ones).

/// What:     `use std::path::{Component, Path, PathBuf};`. `PathBuf` is the OWNED,
///           growable path (sibling: the borrowed `&Path`); `Path` is the borrowed
///           view; `Component` is one piece of a split path (a root `/`, a `..`, a
///           normal name segment, etc.).
/// Why:      We borrow each track path (`&Path`), split it into `Component`s, and
///           keep only the named segments; the input list is owned `PathBuf`s.
///
/// In TS you'd write (pseudocode):
/// ```ts
/// // all three are just strings in TS; Component ~ one element of path.split("/")
/// ```
use std::path::{Component, Path, PathBuf};

/// What:     `const SEPARATOR: &str = "/";`. `&str` is a BORROWED string slice
///           (sibling: the owned `String`); the separator we re-join segments with.
/// Why:      The queue is Linux-only and the UI/pagination expect a single `/`
///           separator; naming it avoids a bare `"/"` literal scattered around.
///
/// In TS you'd write (pseudocode):
/// ```ts
/// const SEPARATOR = "/";
/// ```
const SEPARATOR: &str = "/";

/// What:     `fn normal_components(path: &Path) -> Vec<String>`. Split a path into its
///           NAMED segments only, each as an owned `String`. `&Path` is a read-only
///           borrow; `Vec<String>` is the owned, growable array of owned strings
///           (sibling: the borrowed slice `&[String]`). Private helper.
/// Why:      Dropping the root (`/`) and any `.`/`..` leaves just the folder and file
///           names, so prefix comparison and re-joining are clean and behave the same
///           for absolute and relative inputs.
///
/// In TS you'd write (pseudocode):
/// ```ts
/// function normalComponents(path: string): string[] {
///   return path.split("/").filter(seg => seg !== "" && seg !== "." && seg !== "..");
/// }
/// ```
fn normal_components(path: &Path) -> Vec<String> {
    // What:     `let mut out: Vec<String> = Vec::new();`. `Vec::new()` builds a fresh
    //           empty owned array. `mut` marks it mutable (bindings are read-only by
    //           default); the explicit type is needed because it starts empty.
    // Why:      Accumulate the named segments.
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // const out: string[] = [];
    // ```
    let mut out: Vec<String> = Vec::new();
    // What:     `for component in path.components() { ... }`. `path.components()` is an
    //           iterator over the path's pieces as `Component` values; `for ... in`
    //           consumes it one piece at a time.
    // Why:      Inspect each piece and keep only the named ones.
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // for (const component of path.split("/")) { ... }
    // ```
    for component in path.components() {
        // What:     `if let Component::Normal(part) = component { ... }`. A one-arm
        //           pattern match: run the block only when the piece is the `Normal`
        //           variant (a real name segment, not `/`, `.`, or `..`), binding its
        //           inner `&OsStr` (an OS string slice) to `part`.
        // Why:      Skip the root and relative markers; keep only folder/file names.
        //
        // In TS you'd write (pseudocode):
        // ```ts
        // if (segment !== "" && segment !== "." && segment !== "..") { ... }
        // ```
        if let Component::Normal(part) = component {
            // What:     `out.push(part.to_string_lossy().into_owned());`.
            //           `part.to_string_lossy()` makes a `Cow<str>` (borrowed-or-owned,
            //           replacing invalid UTF-8 bytes); `.into_owned()` forces an owned
            //           `String`; `.push(...)` appends it (MOVING it into the vector).
            // Why:      Store an owned copy that outlives the borrowed path.
            // Gotcha:   `to_string_lossy` may REPLACE invalid bytes with U+FFFD rather
            //           than fail; an OS path is not guaranteed valid UTF-8.
            //
            // In TS you'd write (pseudocode):
            // ```ts
            // out.push(segment);
            // ```
            out.push(part.to_string_lossy().into_owned());
        }
    }
    // What:     `out`. Tail expression (no trailing `;`) -> the function's return.
    // Why:      Hand back the named segments.
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // return out;
    // ```
    out
}

/// What:     `fn common_prefix_len(lists: &[Vec<String>]) -> usize`. How many LEADING
///           segments every track shares, capped so at least one segment (the
///           filename) always remains. `&[Vec<String>]` is a borrowed slice of
///           segment lists; `usize` is the pointer-sized unsigned integer used for
///           lengths/indices (siblings: `u32`, `u64`, `i32`). Private helper.
/// Why:      That shared prefix is the "loaded root"; stripping it turns absolute
///           paths into the relative paths the UI shows.
///
/// In TS you'd write (pseudocode):
/// ```ts
/// function commonPrefixLen(lists: string[][]): number { ... }
/// ```
fn common_prefix_len(lists: &[Vec<String>]) -> usize {
    // What:     `let shortest = lists.iter().map(|list| list.len()).min().unwrap_or(0);`.
    //           `.iter()` borrows each list; `.map(|list| list.len())` turns each into
    //           its length; `.min()` returns `Option<usize>` (the smallest, or `None`
    //           for no lists); `.unwrap_or(0)` extracts it or substitutes 0. `|list|
    //           ...` is a closure taking a borrowed list.
    // Why:      The common prefix cannot be longer than the shortest path.
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // const shortest = lists.length ? Math.min(...lists.map((l) => l.length)) : 0;
    // ```
    let shortest = lists.iter().map(|list| list.len()).min().unwrap_or(0);
    // What:     `if shortest == 0 { return 0; }`. Early return when some path has no
    //           named segments (nothing to strip).
    // Why:      Avoid the `shortest - 1` underflow below (`usize` is unsigned, so
    //           `0 - 1` would panic in debug builds).
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // if (shortest === 0) return 0;
    // ```
    if shortest == 0 {
        return 0;
    }
    // What:     `let cap = shortest - 1;`. The most segments we may strip, leaving at
    //           least the final one (the filename) on every track.
    // Why:      A row must never collapse to an empty label.
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // const cap = shortest - 1;
    // ```
    let cap = shortest - 1;
    // What:     `let mut run: usize = 0;`. A counter for how many leading segments
    //           match so far. `mut` because the loop increments it.
    // Why:      Count the shared run.
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // let run = 0;
    // ```
    let mut run: usize = 0;
    // What:     `while run < cap && lists.iter().all(|list| list[run] == lists[0][run]) { ... }`.
    //           Keep going while we are under the cap AND every list agrees on the
    //           segment at position `run`. `.all(|list| ...)` is true only when the
    //           closure holds for every element; `list[run]` indexes a list (safe:
    //           `run < cap < shortest <= every length`). `lists[0]` is the reference
    //           list. `&&` short-circuits, so `.all` runs only while under the cap.
    // Why:      One linear scan finds the longest shared leading run, no recursion.
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // while (run < cap && lists.every((l) => l[run] === lists[0][run])) run++;
    // ```
    while run < cap && lists.iter().all(|list| list[run] == lists[0][run]) {
        // What:     `run += 1;`. Advance past a matching segment.
        // Why:      Move to the next position.
        //
        // In TS you'd write (pseudocode):
        // ```ts
        // run++;
        // ```
        run += 1;
    }
    // What:     `run`. Tail expression -> the shared-prefix length.
    // Why:      Hand back how many leading segments to strip.
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // return run;
    // ```
    run
}

/// What:     `pub fn relative_display_paths(tracks: &[PathBuf]) -> Vec<String>`. Turn
///           each track's full path into its path relative to the queue's common
///           root. `&[PathBuf]` is a borrowed slice of owned paths (read-only); the
///           result is one owned relative string per track, in order.
/// Why:      The UI shows folders, not just filenames, but the absolute prefix (e.g.
///           `/home/user/Music`) is noise; this strips it once per queue.
///
/// In TS you'd write (pseudocode):
/// ```ts
/// function relativeDisplayPaths(tracks: string[]): string[] {
///   if (tracks.length === 0) return [];
///   const lists = tracks.map(normalComponents);
///   const prefix = commonPrefixLen(lists);
///   return lists.map((list, i) => {
///     const rel = list.slice(prefix).join("/");
///     return rel === "" ? tracks[i] : rel;
///   });
/// }
/// ```
pub fn relative_display_paths(tracks: &[PathBuf]) -> Vec<String> {
    // What:     `if tracks.is_empty() { return Vec::new(); }`. Early return for an
    //           empty queue. `Vec::new()` builds an empty owned array.
    // Why:      Nothing to relativize; `common_prefix_len` of nothing is undefined.
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // if (tracks.length === 0) return [];
    // ```
    if tracks.is_empty() {
        return Vec::new();
    }
    // What:     `let component_lists: Vec<Vec<String>> = tracks.iter().map(|p| normal_components(p)).collect();`.
    //           Borrow each track (`p: &PathBuf`, which derefs to the `&Path`
    //           `normal_components` takes), split it into named segments, and
    //           `.collect()` the per-track lists into one vector of lists.
    // Why:      Compute every path's segments once, reused for the prefix and the
    //           per-track remainder.
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // const componentLists = tracks.map(normalComponents);
    // ```
    let component_lists: Vec<Vec<String>> = tracks.iter().map(|p| normal_components(p)).collect();
    // What:     `let prefix_len = common_prefix_len(&component_lists);`. `&...` lends
    //           the lists read-only to the helper.
    // Why:      Decide how many leading segments are the shared root.
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // const prefixLen = commonPrefixLen(componentLists);
    // ```
    let prefix_len = common_prefix_len(&component_lists);
    // What:     `component_lists.iter().zip(tracks.iter()).map(|(list, path)| { ... }).collect()`.
    //           `.zip(tracks.iter())` pairs each segment list with its original path
    //           so the fallback can reach the full path; `.map(|(list, path)| ...)`
    //           takes a DESTRUCTURED tuple of the two and builds each relative string;
    //           `.collect()` gathers them into the returned `Vec<String>`. Tail
    //           expression -> return.
    // Why:      Produce one relative path per track, preserving load order.
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // return componentLists.map((list, i) => { ... });
    // ```
    component_lists
        .iter()
        .zip(tracks.iter())
        .map(|(list, path)| {
            // What:     `let relative = list[prefix_len..].join(SEPARATOR);`.
            //           `list[prefix_len..]` is a slice of the segments AFTER the
            //           shared prefix (a RANGE index; `..` means "to the end");
            //           `.join(SEPARATOR)` glues them with `/` into an owned `String`.
            // Why:      The segments past the common root ARE the relative path.
            //
            // In TS you'd write (pseudocode):
            // ```ts
            // const relative = list.slice(prefixLen).join("/");
            // ```
            let relative = list[prefix_len..].join(SEPARATOR);
            // What:     `if relative.is_empty() { path.to_string_lossy().into_owned() } else { relative }`.
            //           An `if/else` EXPRESSION (no `;`) evaluating to one branch's
            //           `String`. `path.to_string_lossy().into_owned()` is the
            //           full-path fallback (lossy UTF-8, then owned).
            // Why:      Defensive: a pathological path with no named segments would
            //           join to "", so show its full text instead of nothing.
            //
            // In TS you'd write (pseudocode):
            // ```ts
            // return relative === "" ? path : relative;
            // ```
            if relative.is_empty() {
                path.to_string_lossy().into_owned()
            } else {
                relative
            }
        })
        .collect()
}

/// What:     `#[cfg(test)] #[path = "relpath_tests.rs"] mod tests;` declares a
///           test-only submodule whose code lives in the sibling file
///           `relpath_tests.rs`. `#[cfg(test)]` gates it to test builds only;
///           `#[path = "..."]` aims the module at a flat sibling file instead of the
///           default `relpath/tests.rs` subdirectory lookup. The file stays the
///           `tests` CHILD of relpath, so its `use super::*` reaches the module items
///           (including private ones) unchanged.
/// Why:      Keep `relpath.rs` to production code; the tests live beside it without
///           inflating this file or its max-lines budget (sibling `*_tests.rs` files
///           are exempt from the linter).
///
/// In TS you'd write (pseudocode):
/// ```ts
/// // relpath.unit.test.ts, run only by the test runner
/// ```
#[cfg(test)]
#[path = "relpath_tests.rs"]
mod tests;
